simple_certmanager/forms.py

```diff
@@ -52,6 +67,46 @@ def save(self, commit=True):
             instance.save()
         return instance
 
+    def clean(self) -> dict[str, Any]:
+        # Errors if the CSR already has a certificate
+        if self.instance.public_certificate:
+            self.add_error(
+                "certificate",
+                "A certificate already exists for this CSR. "
+                "Delete the certificate first.",
+            )
+        # Handle the certificate file upload
+        if "certificate" in self.cleaned_data and self.cleaned_data["certificate"]:
+            uploaded_file = self.cleaned_data["certificate"]
+            try:
+                uploaded_file.seek(0)
+                certificate_data = uploaded_file.read()
+
+                # Load the certificate and throw an error if the certificate is invalid
+                certificate = x509.load_pem_x509_certificate(certificate_data)
+                private_key = load_pem_x509_private_key(
+                    self.instance.private_key.encode()
+                )
```

Comment:
I think everywhere else we explicitly encode as `ascii` - please don't rely on implicit encoding parameters when converting to/from binary data since it can lead to bugs that are hard to track down.

I believe that the Python 2 -> Python 3 transition changed the default value of the encoding, so making it explicit is much more future proof.